sql/src/main/java/org/apache/druid/sql/calcite/rel/VirtualColumnRegistry.java

```diff
@@ -32,52 +32,38 @@
 import java.util.TreeSet;
 
 /**
- * Wraps a {@link RowSignature} and provides facilities to re-use {@link VirtualColumn} definitions for dimensions,
- * filters, and filtered aggregators while constructing a {@link DruidQuery}
+ * Provides facilities to create and re-use {@link VirtualColumn} definitions for dimensions, filters, and filtered
+ * aggregators while constructing a {@link DruidQuery}.
  */
-public class DruidQuerySignature
+public class VirtualColumnRegistry
```

Comment:
The point in this PR where I started splitting DruidQuerySignature into RowSignature + VirtualColumnRegistry was when writing this code:

```java
      this.sorting = Preconditions.checkNotNull(
          computeSorting(
              partialQuery,
              plannerContext,
              computeOutputRowSignature(),
              // When sorting follows grouping, virtual columns cannot be used
              partialQuery.getAggregate() != null ? null : virtualColumnRegistry
          )
      );
```

To get the proper DruidQuerySignature without the split would have meant doing:

```java
      RowSignature signature = computeOutputRowSignature();
      this.sorting = Preconditions.checkNotNull(
          computeSorting(
              partialQuery,
              plannerContext,
              // When sorting follows grouping, virtual columns cannot be used
              partialQuery.getAggregate() != null
              ? new DruidQuerySignature(this.grouping.getOutputRowSignature()).asAggregateSignature(this.grouping.getOutputRowSignature())
              : notSureWhatGoesHere
          )
      );
```

`notSureWhatGoesHere` reflects this situation:

- At this point, the sort might be after a select or a select + project.
- If the sort is after a simple select, it should be `sourceQuerySignature`.
- If the sort is after a select + project, then the sort refers to the project (i.e. the `selectProjection`) which changed the row signature, and might have created its own virtual columns. Passing in `sourceQuerySignature` is wrong, since it's the wrong starting signature (it's pre-select-project). But passing in a new DruidQuerySignature is wrong too, since it will 'forget' about virtual columns that might have been used by the selectProjection.

The solution seems to be to split it, so we can pass in `computeOutputRowSignature()`, the new post-selectProjection signature, along with the original `virtualColumnRegistry`.

The change also made `computeHavingFilter` nicer, since it would take a RowSignature instead of DruidQuerySignature. (That was the only place where asAggregateSignature was used.)

I thought it also made it clearer in various places what was going on, if a bit more verbose.

Reply:
Yeah I agree the change makes more sense in this world that i think also more accurately reflects reality, but where the stages are less obviously split into 'pre' and 'post' stuff, like I said not disputing the changes here 😅. It seems like it would be a bit messy in a couple other places too if you had to treat it as potentially in either 'mutable' or 'immutable' context depending on the type query, so I'm on board. 

I ever find the spare time, I will still probably look into if it's possible to consolidate the agg re-use boilerplate code into what is now `VirtualColumnRegistry` to help simplify sql agg implementations where possible though.